client/client.go

```diff
@@ -403,26 +403,6 @@ type response struct {
 	Maintenance *Error `json:"maintenance"`
 }
 
-// Error is the real value of response.Result when an error occurs.
-type Error struct {
-	Kind    string      `json:"kind"`
-	Value   interface{} `json:"value"`
-	Message string      `json:"message"`
-
-	StatusCode int
```

Comment:
I have looked at snapd and pebble, and I cannot actually find this being used. I've built snapd + unit tests, and pebble + unit tests without this line, and all works. I propose we drop this until a real need arises. The Status code and test version is available in the ```response struct```, of which this ```Error``` is unmarshalled from ```response.Result```.

Reply:
Yeah, seems okay to remove, but I'd argue that it's for a different reason: the key reason why supposedly it's helpful to have this code here is because it would be nice to know that the server is rejecting our request based on external factors (e.g. it's a proxy, and it doesn't want to talk to us at all).  Ironically, our code currently only sets the status code when it detects that the response unmarshals an error type properly, and by then we're already past the point of caring about the status code.  That's why it seems irrelevant to have the status code here, at least as the logic stands, and it also explains why no code is using it. So +1 on dropping it for now.

Reply:
Actually, we may need to roll this back. It looks like the status code is being intentionally used already in some endpoints, and is part of tests too:

https://github.com/canonical/pebble/pull/299#discussion_r1327321448

Reply:
Hi Gustavo. I am not immediately sure how to best improve this:

[rsp.err](https://github.com/canonical/pebble/blob/cb90b548699363124ad7b009cecb2c3bfc301457/client/client.go#L426), as you know, currently discards the status code of any reponse type apart from type ```error```. This means the ```health``` [test](https://github.com/canonical/pebble/pull/299/files#r1328310832) code that sets the status code on reply for a ```sync``` type, will not result in an error. The status code currently is discarded.

Reply:
In the client responder interface, I am offering the Status Code now as part of the common  return (as you have seen). Could this provide the plumbing for code to get access to the status code consistently, instead of through the error ?

```go
// DecoderFunc allows the client access to the HTTP response. See the SetDecoder
// description for more details.
type DecoderFunc func(ctx context.Context, res *http.Response, opts *RequestOptions, result interface{}) (*RequestResponse, error)

type Requester interface {
	// Do must support the following cases:
	//
	// 1. Sync response
	// 2. Async response
	// 3. Error response
	// 4. Websocket creation request
	// 5. Raw HTTP body request
	//
	// See the default implementation for information on what is expected in
	// each use case.
	Do(ctx context.Context, opts *RequestOptions, result interface{}) (*RequestResponse, error)

	// SetDecoder allows for client specific processing to be hooked into
	// the sync and async response decoding process. The decoder is also
	// responsible for unmarshalling the result, and populating the
	// RequestReponse.
	SetDecoder(decoder DecoderFunc)
}

// RequestOptions allows setting up a specific request.
type RequestOptions struct {
	Method  string
	Path    string
	Query   url.Values
	Headers map[string]string
	Body    io.Reader
	Async   bool
}

// RequestResponse defines a common response associated with requests.
type RequestResponse struct {
	StatusCode int
	Type       string
	Change     string
}
```

Reply:
I have a MM agreement with Gustavo that I will remove StatusCode from the Error type, but we will consistently expose it through the new interface planned in the next PR, as hinted above.

Reply:
Hmm, I don't like this removal of `StatusCode` at all -- this is an exported field of a public package, so I think we're playing fast and loose with backwards-compatibility. It *might* be safe to remove in this instance (we don't know), but as Pebble and the `pebble/client` package become more widely used, it's likely that someone is using that field.

I'm also not sure that @niemeyer's statement is true: "That's why it seems irrelevant to have the status code here, at least as the logic stands", because if there's an error you might want to drill down on the response code. For example, a package user could be calling `client.Change(id)`, and on error, type-asserting to `*client.Error` and checking whether the status code is 404 (change ID not found).

In this case the daemon returns an `Error.Kind` of `""`, so you can't test that. (This should probably be considered a bug -- I think the server should set `Kind` to `ErrorKindNotFound`, but it currently does not.)

Example (I've tested this locally and it correctly distinguishes between not-found and other server error codes correctly):

```go
	chg, err := cli.Change(id)
	if err != nil {
		clientErr, ok := err.(*client.Error)
		if ok && clientErr.StatusCode == 404 {
			// Handle not-found specially
			fmt.Printf("change ID %q not found!\n", chid)
		}
		return err
	}
```

By removing `Error.StatusCode`, this code would break. I agree it's somewhat unlikely this is being used (and clearly not in snapd or Pebble). But I don't think we can assume we don't have other `pebble/client` package users.

Reply:
@benhoyt @niemeyer I think I successfully mislead Gustavo in the claim that a Sync or ASync reponse can only be of type Sync or Async ... I see your point. The ```resp.err(...)``` happens before the check for type, and therefore allows such a call to return an error type! Glad you caught this.